assets/js/wpjm-stats.js

```diff
@@ -0,0 +1,78 @@
+/* global job_manager_stats */
+( function () {
+	// From https://youmightnotneedjquery.com/#ready
+	function ready( fn ) {
+		if ( document.readyState !== 'loading' ) {
+			fn();
+		} else {
+			document.addEventListener( 'DOMContentLoaded', fn );
+		}
+	}
+
+	// Cookie funcs are copied verbatim from http://www.quirksmode.org/js/cookies.html, tweaked to include path.
+	function createCookie( name, value, days, path ) {
+		var expires = '';
+		if ( days ) {
+			var date = new Date();
+			date.setTime( date.getTime() + days * 24 * 60 * 60 * 1000 );
+			var expires = '; expires=' + date.toGMTString();
+		} else var expires = '';
+		document.cookie = name + '=' + value + expires + '; path=' + path;
+	}
+
+	function readCookie( name ) {
+		var nameEQ = name + '=';
+		var ca = document.cookie.split( ';' );
+		for ( var i = 0; i < ca.length; i++ ) {
+			var c = ca[ i ];
+			while ( c.charAt( 0 ) == ' ' ) c = c.substring( 1, c.length );
+			if ( c.indexOf( nameEQ ) == 0 ) return c.substring( nameEQ.length, c.length );
+		}
+		return null;
+	}
+
+	function eraseCookie( name ) {
+		createCookie( name, '', -1 );
+	}
+
+	ready( function () {
+		var statsToRecord = [];
+		var jobStatsSettings = window.job_manager_stats;
+		var ajaxUrl = jobStatsSettings.ajax_url;
+		var ajaxNonce = jobStatsSettings.ajax_nonce;
+		var cookiesToSet = [];
+		var setCookies = function () {
+			cookiesToSet.forEach( function ( uniqueKey ) {
+				createCookie( uniqueKey, 1, 1, '' );
+			} );
+		};
+
+		jobStatsSettings.stats_to_log.forEach( function ( statToRecord ) {
```

Comment:
I'd include the stat name in the payload instead of having a fixed list — that'd be hard to scale to other kind of events.

Thinking about how to architect this all, we could add an ajax handler to process the request, and then pass on the event to specialized handlers. Looks best to start separating some parts out to a new `Job_Listing_Stats` class to make reasoning about this easier. 

On the JS side, we'd want to expose some global public `log_stats` function that only handles sending the event to the server. Then we have various entry points that'd call this:

- After page load, a function logs the `job_listing_view` and `job_listing_view_unique` events, depending on cookies/storage. Or with the schema explored below, it calls the `job_listing_view` handler with `$unique` as a parameter, and let the server decide.
- On clicking 'Apply Now', it calls the `job_listing_apply_click` handler

Here is a sketch for the backend parts. I really like the idea of loosely coupling all this with the WordPress hooks approach:

```php

class Stats {

	// Option one, register callbacks for each
	public function get_registered_stats() {
		return apply_filters('wpjm_get_registered_stats', [
			'job_listing_view'  => [ Job_Listing_Stats::class, 'log_view' ],
			'job_listing_apply' => [ Job_Listing_Stats::class, 'log_apply_click' ],
		] );
	}

	public function maybe_log_stat_ajax() {
		
		// process request & everything

		call_user_func( $this->get_registered_stats()[ $params[ 'stat' ] ], $params );

		// The WordPress hooks option
		do_action( "job_manager_stat__{$params[ 'stat' ]}" )

	}
}

class Job_Listing_Stats {

	// Alternatively option, using WordPress hooks
	public function __construct() {
		add_action("job_manager_stat__job_listing_view", [ $this, 'log_view' ] );
		add_action("job_manager_stat__job_listing_apply", [ $this, 'log_apply' ] );
	}

	public function log_view( $post_id, $unique = false ) {

		// checks and whistles

		$stats->log_stat("job_listing_view", $post_id )

		if $unique log_stat("job_listing_view_unique", $post_id )
	}
}
```

Reply:
I think I like the first version, and steered the pr to that direction. I will leave things in the same class for now and we can update/re-architect as we implement more stats use cases.